server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+                               LongSupplier currentTimeSupplier) {
+            super(settings, publishRequest, ackListener, currentTimeSupplier);
+            this.publishRequest = publishRequest;
+        }
+
+        @Override
+        protected void onCompletion(boolean committed) {
+            completed = true;
+            this.committed = committed;
+        }
+
+        @Override
+        protected void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response) {
+
+        }
+
+        @Override
+        protected void sendPublishRequest(DiscoveryNode destination, PublishRequest publishRequest,
+                                          ActionListener<PublishWithJoinResponse> responseActionListener) {
+            assertSame(publishRequest, this.publishRequest);
+            assertNull(pendingPublications.put(destination, responseActionListener));
+        }
+
+        @Override
+        protected void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommit,
+                                       ActionListener<TransportResponse.Empty> responseActionListener) {
+            if (this.applyCommit == null) {
+                this.applyCommit = applyCommit;
+            } else {
+                assertSame(applyCommit, this.applyCommit);
+            }
+            assertNull(pendingCommits.put(destination, responseActionListener));
+        }
+    }
+
+    DiscoveryNode n1 = CoordinationStateTests.createNode("node1");
+    DiscoveryNode n2 = CoordinationStateTests.createNode("node2");
+    DiscoveryNode n3 = CoordinationStateTests.createNode("node3");
+    Set<DiscoveryNode> discoNodes = Sets.newHashSet(n1, n2, n3);
+
+    MockNode node1 = new MockNode(Settings.EMPTY, n1);
+    MockNode node2 = new MockNode(Settings.EMPTY, n2);
+    MockNode node3 = new MockNode(Settings.EMPTY, n3);
+    List<MockNode> nodes = Arrays.asList(node1, node2, node3);
+
+    Function<DiscoveryNode, MockNode> nodeResolver = dn -> nodes.stream().filter(mn -> mn.localNode.equals(dn)).findFirst().get();
+
+    private void initializeCluster(VotingConfiguration initialConfig) {
+        node1.coordinationState.setInitialState(CoordinationStateTests.clusterState(0L, 1L, n1, initialConfig, initialConfig, 0L));
+        StartJoinRequest startJoinRequest = new StartJoinRequest(n1, 1L);
+        node1.coordinationState.handleJoin(node1.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node2.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node3.coordinationState.handleStartJoin(startJoinRequest));
+        assertTrue(node1.coordinationState.electionWon());
+    }
+
+    public void testSimpleClusterStatePublishing() throws InterruptedException {
+        VotingConfiguration singleNodeConfig = new VotingConfiguration(Sets.newHashSet(n1.getId()));
```

Comment:
Could we also have a demonstration of the happy path on a three-node configuration, with the assertions adjusted accordingly? In the three-node case it's possible that publication responses interleave with commits, and this should be covered.

Reply:
Even in a single-node configuration cluster, publish responses can arrive after commit responses. We already test here that publish responses arrive before committing and after committing. We don't test commit responses arriving before all publish responses are handled. I've added this in  0199b3de7b2